### PR TITLE
Fix #4191 make websockets work in angular in parallel with http

### LINKED ISSUE #4191: Websocket communication to client
See #4177.

Develop message protocol over websockets that can be deployed in the supervisor as tornado server, running inside the supervisor so that we can seamlessly support new clients. The protocol will work with a simple client that can render the progress of a simulation. This is not about the data model or UI. Rather, just the communication between client and server.

### LINKED ISSUE #6119: change srunit to use websockets when enabled
Simulate what the JS UI does when making requests: if authenticated, start a websocket and communicate over that.

### LINKED ISSUE #6205: authState redirects to root if user dir not found
authState should always return js, because it is embedded in js. There should never be an HTML return or a redirect.

### PR DESCRIPTION
**Pull request by robnagler**
TODO:
- [x] http prefix 
- [x] promoting to email user (need to use $http)
- [x] Fix #6205 authState exception
- [x] Fix #6119 websockets in srunit

Depends on radiasoft/pykern#388

### CODE REVIEW

File: sirepo/uri_router.py [hunk 1]
```diff
@@ -218,11 +224,74 @@ async def get(self):
...
+                "start ws={} msg={} remote_ip={}",
+                self.ws_id,
+                self.msg_count,
+                self.remote_ip,
+            )
+            # TODO(robnagler) what if msg poorly constructed? Close socket?
+            c = pkjson.load_any(msg)
+            e, r, k = _path_to_route(c.uri[1:])
+            if e:
+                pkdlog("uri={} {}; route={} kwargs={} ", c.uri, e, r, k)
```
Comment (e-carlin) [inline sirepo/uri_router.py:244]: Maybe add a pkdc controlled dump of the full message contents? I frequently use the network tab in the browser to see what payloads are going back and forth which no longer works with ws.

File: sirepo/request.py [hunk 2]
```diff
@@ -30,7 +31,7 @@ def init_quest(qcall, internal_req=None):
             http_authorization=None,
             http_headers=PKDict(),
             http_method="GET",
-            http_request_uri="/",
+            #            http_request_uri="/",
```
Comment (e-carlin) [inline sirepo/request.py:34]: odd formatting?

**Review by e-carlin — CHANGES_REQUESTED** [review]
Code all looks good. The browser seems to be faster \<snap\>. 

I noticed a few issues. Not positive they are related to this work:
1. srw -> light source facilities -> nsls2 -> nsls2 esm beam line -> nsls2 elm beam line -> coherent results
```
sirepo/job_driver/__init__.py:306:_receive LocalDriver(a=k5Ek k=sequential u=Ir6N [_Op(analysis, DcBO)]) error msg={'agentId': 'k5EksGpsROcOy0zQGc0Qh6F43lBwJFyF', 'opId': 'DcBOhnM0nB7ek48csrWybOJ10RmhO5U5', 'opName': 'analysis', 'reply': {'error': 'report not generated', 'stack': 'Traceback (most re\
cent call last):
  File "/home/vagrant/src/radiasoft/sirepo/sirepo/pkcli/job_cmd.py", line 252, in _do_get_simulation_frame
    return template_common.sim_frame_dispatch(
  File "/home/vagrant/src/radiasoft/sirepo/sirepo/template/template_common.py", line 705, in sim_frame_dispatch
    res = o(frame_args)
  File "/home/vagrant/src/radiasoft/sirepo/sirepo/template/srw.py", line 507, in sim_frame
    m = _copy_frame_args_into_model(frame_args, "watchpointReport")
  File "/home/vagrant/src/radiasoft/sirepo/sirepo/template/srw.py", line 483, in _copy_frame_args_into_model
    m[f] = float(m[f])
ValueError: could not convert string to float: '1e%2B30'
```
2. Url for Jupyter aren't correct (ex http://localhost/jupyterhublogin instead of http://127.0.0.1:8080/jupyterhublogin). This is when running with etc/run-jupyterhub.sh
3. srw esm beamline when I clicked on "Open as new shadow simulation" and then clicked on the provided link
```
angular.js:15567 TypeError: Cannot read properties of undefined (reading 'indexOf')
    at iconUrl (sirepo-beamline.js?20230718.161240:480:26)
    at controller (sirepo-beamline.js?20230718.161240:505:30)
    at Object.invoke (angular.js:5141:16)
    at S.instance (angular.js:11704:24)
    at p (angular.js:10517:33)
    at g (angular.js:9832:13)
    at g (angular.js:9835:13)
    at p (angular.js:10579:11)
    at g (angular.js:9832:13)
    at g (angular.js:9835:13)
```

**Comment by robnagler** [comment]
- working end-to-end with limited browser testing
- half-integrated msgpack (binary messages) which was necessary to get srw image sample preview working
- unit tests use websockets like the browser (if feature_config.ui_websocket enabled)
- auth still occurs through http
- http prefix is still not coded, but not necessary for most functions (just jupyter)

Next:
- http prefix
- msgpack for requests
- use msgpack, not json for dict reply/request

**Comment by robnagler** [comment]
@e-carlin this is ready for another review. I think it could be merged (only on in dev). The next big issues are caching frames and pub/sub for runStatus. These should be separate concerns. The former is necessary for a prod release but not the latter.

One question I have is how frequently should the UI try restarts on the websocket. I think polling is fine, since it's a lightweight request. There might need to be a user alert in the UI that connection has been interrupted from more than, say, 30 seconds (server restart).

**Review by e-carlin — APPROVED** [review]
Only one small comment on a comment.

All of my click around testing worked. Obviously there are counltess cases I didn't manually test so it would be good for other people to click around edge cases in their apps.

One general comment, I find myself frequently opening the developer tools network tab to view what requests are made and what the content of those requests are. With websockets and a binary protocol that now doesn't work. What about adding a pkdc to parse_message and websocket_response that allows one to dump the message (with associated uri, seq num, etc)?

File: sirepo/package_data/static/js/sirepo.js [hunk 3]
```diff
@@ -2004,15 +2018,232 @@ SIREPO.app.factory('panelState', function(appState, requestSender, simulationQue
...
+        let i = 0;
+        for (const b of buffers) {
+            f.set(b, i);
+            i += b.length;
+        }
+        wsreq.msg = f;
+    };
+
+    const _error = (event) => {
+        // close: event.code : short, event.reason : str, wasClean : bool
```
[the remark this answers is not in the document]
Comment (robnagler) [reply sirepo/package_data/static/js/sirepo.js:2041]: These are standard codes in JS: https://developer.mozilla.org/en-US/docs/Web/API/WebSocket/close_event

File: sirepo/package_data/static/js/sirepo.js [hunk 4]
```diff
@@ -1985,15 +1979,178 @@ SIREPO.app.factory('panelState', function(appState, requestSender, simulationQue
...
-SIREPO.app.factory('requestSender', function(cookieService, errorService, utilities, $http, $location, $injector, $interval, $q, $rootScope, $window) {
+// cannot import authState factory, because of circular import with requestSender
+SIREPO.app.factory('msgRouter', ($http, $interval, $q, $window, errorService) => {
+    const self = {};
+    const toSend = [];
+    let needReply = {};
+    let reqSeq = 1;
+    let socket = null;
+
+    const _isAuthenticated = () =>  {
```
[the remark this answers is not in the document]
Comment (robnagler) [reply sirepo/package_data/static/js/sirepo.js]: Fixed

**Comment by robnagler** [comment]
https://github.com/radiasoft/sirepo/issues/6250
